### PR TITLE
Supershamir refactor

### PR DESCRIPTION
**Pull request by ciny**
This PR will implement the epic tracked in #442

### CODE REVIEW

File: tests/device_tests/common.py [hunk 1]
```diff
@@ -30,7 +30,6 @@
     "extra extend academic arcade born dive legal hush gross briefing talent drug much home firefly toxic analysis idea umbrella slice",
 ]
 MNEMONIC_SHAMIR_20_2of3_2of3_GROUPS = [
-    "gesture negative ceramic leaf device fantasy style ceramic safari keyboard thumb total smug cage plunge aunt favorite lizard intend peanut",
```
Comment (matejcik) [inline tests/device_tests/common.py]: why is this removed?

Comment (ciny) [reply tests/device_tests/common.py]: It is an extra share that we use to test that we don't pass unfinished groups to crypto.slip39 during recovery. It was moved to EXTRA_GROUP_SHARE in the relevant test (recoverydevice_supershamir). IIRC the reason for the move was that load_device/setup_client pass everything directly to crypto.slip39 and adding an extra share where it is needed is more obvious than cutting up a list imported from common every time we want to use it.

File: core/src/apps/management/reset_device/layout.py [hunk 2]
```diff
@@ -208,113 +278,70 @@
...
-    checklist.add("Set group threshold")
-    checklist.add(("Set size and threshold", "for each group"))
-    checklist.select(0)
-    return await confirm(
-        ctx, checklist, ButtonRequestType.ResetDevice, cancel=None, confirm="Continue"
-    )
-
+async def slip39_prompt_threshold(ctx, num_of_shares, group_id=None):
+    count = num_of_shares // 2 + 1
+    min_count = min(2, num_of_shares)
```
Comment (matejcik) [inline core/src/apps/management/reset_device/layout.py]: The `min` is weird. It is always 2, unless `num_of_shares` is 1? Is this the thing with the 1-of-1 share? If so, an explanatory comment would help.

File: core/src/apps/management/reset_device/__init__.py [hunk 3]
```diff
@@ -1,34 +1,31 @@
...
 
 
 async def reset_device(ctx: wire.Context, msg: ResetDevice) -> Success:
     # validate parameters and device state
     _validate_reset_device(msg)
 
-    is_slip39_simple = msg.backup_type == ResetDeviceBackupType.Slip39_Single_Group
-    is_slip39_group = msg.backup_type == ResetDeviceBackupType.Slip39_Multiple_Groups
-
     # make sure user knows he's setting up a new wallet
```
[the remark this answers is not in the document]
Comment (tsusanka) [reply core/src/apps/management/reset_device/__init__.py]: Fixed

File: tests/device_tests/test_msg_recoverydevice_shamir_dryrun.py [hunk 4]
```diff
@@ -54,7 +54,7 @@ def input_flow():
             label="label",
             language="english",
             dry_run=True,
-            type=messages.ResetDeviceBackupType.Slip39_Single_Group,
+            type=messages.BackupType.Slip39_Basic,
```
Comment (matejcik) [inline tests/device_tests/test_msg_recoverydevice_shamir_dryrun.py]: this is actually an inappropriate value, `type` is "matrix" or "scrambled words", only relevant on T1. so please remove this line

File: tests/device_tests/test_msg_recoverydevice_shamir_dryrun.py [hunk 5]
```diff
@@ -85,5 +85,5 @@ def input_flow():
...
```
Comment (matejcik) [inline tests/device_tests/test_msg_recoverydevice_shamir_dryrun.py]: dtto

File: core/src/apps/management/recovery_device/homescreen.py [hunk 6]
```diff
@@ -134,93 +172,79 @@
...
             raise RuntimeError
 
         if group_count > 1:
-            content = layout.RecoveryHomescreen(
-                "More shares needed", "for this recovery"
-            )
+            content = layout.RecoveryHomescreen("More shares needed")
             await layout.homescreen_dialog(ctx, content, "Enter share")
         else:
             if remaining[0] == 1:
```
[the remark this answers is not in the document]
Comment (matejcik) [reply core/src/apps/management/recovery_device/homescreen.py]: we don't because it is waiting for UI tests

**Comment by matejcik** [comment]
let's not forget about these:
https://github.com/trezor/trezor-firmware/commit/81f5cbef930bf62302d33eedbda4e806b1418172#r35073711
https://github.com/trezor/trezor-firmware/commit/81f5cbef930bf62302d33eedbda4e806b1418172#r35073424
https://github.com/trezor/trezor-firmware/commit/81f5cbef930bf62302d33eedbda4e806b1418172#r35073687
https://github.com/trezor/trezor-firmware/commit/81f5cbef930bf62302d33eedbda4e806b1418172#r35073886

**Comment by ciny** [comment]
> let's not forget about these:
> [81f5cbe#r35073711](https://github.com/trezor/trezor-firmware/commit/81f5cbef930bf62302d33eedbda4e806b1418172#r35073711)
> [81f5cbe#r35073424](https://github.com/trezor/trezor-firmware/commit/81f5cbef930bf62302d33eedbda4e806b1418172#r35073424)
> [81f5cbe#r35073687](https://github.com/trezor/trezor-firmware/commit/81f5cbef930bf62302d33eedbda4e806b1418172#r35073687)
> [81f5cbe#r35073886](https://github.com/trezor/trezor-firmware/commit/81f5cbef930bf62302d33eedbda4e806b1418172#r35073886)

I'll have a look at that.

File: core/src/apps/management/reset_device/__init__.py [hunk 7]
```diff
@@ -157,24 +159,24 @@
...
     if msg.backup_type not in (
-        ResetDeviceBackupType.Bip39,
-        ResetDeviceBackupType.Slip39_Single_Group,
-        ResetDeviceBackupType.Slip39_Multiple_Groups,
+        BackupType.Bip39,
+        BackupType.Slip39_Basic,
+        BackupType.Slip39_Advanced,
     ):
         raise wire.ProcessError("Backup type not implemented.")
     if msg.strength not in (128, 256):
```
[the remark this answers is not in the document]
Comment (tsusanka) [reply core/src/apps/management/reset_device/__init__.py]: Fixed. I made it explicit - will push soon.

File: core/src/apps/management/reset_device/layout.py (hunk 2 above)
Comment (ciny) [reply core/src/apps/management/reset_device/layout.py]: added a comment to clarify

File: core/src/apps/management/recovery_device/layout.py [hunk 8]
```diff
@@ -72,41 +69,60 @@
...
+        # only if we know the backup type we can perform additional checks
+        if backup_type and not await check_mnemonic_validity(ctx, i, word, backup_type):
+            return None
 
         words.append(word)
 
     return " ".join(words)
 
 
+async def check_mnemonic_validity(
```
[the remark this answers is not in the document]
Comment (ciny) [reply core/src/apps/management/recovery_device/layout.py]: renamed to check_slip39_mnemonic... to add clarity

File: core/src/apps/management/reset_device/__init__.py [hunk 9]
```diff
@@ -198,15 +200,13 @@ def _compute_secret_from_entropy(
     return secret
 
 
-async def _show_reset_device_warning(
-    ctx, backup_type: ResetDeviceBackupType = ResetDeviceBackupType.Bip39
-):
+async def _show_reset_device_warning(ctx, backup_type: BackupType = BackupType.Bip39):
```
[the remark this answers is not in the document]
Comment (ciny) [reply core/src/apps/management/reset_device/__init__.py]: fixed

File: core/src/apps/management/reset_device/layout.py [hunk 10]
```diff
@@ -88,7 +159,6 @@
 async def _confirm_word(
     ctx, share_index, numbered_share_words, count, group_index=None
 ):
-    # TODO: duplicated words in the choice list
```
[the remark this answers is not in the document]
Comment (ciny) [reply core/src/apps/management/reset_device/layout.py]: added the TODO back

File: core/src/apps/management/recovery_device/recover.py [hunk 11]
```diff
@@ -24,47 +21,62 @@ def process_bip39(words: str) -> bytes:
...
+            )
+            return secret, share
+        else:
+            # we need more shares
+            return None, share
+
+    if remaining[share.group_index] == 0:
         raise GroupThresholdReachedError()
+
     # These should be checked by UI before so it's a Runtime exception otherwise
```
[the remark this answers is not in the document]
Comment (ciny) [reply core/src/apps/management/recovery_device/recover.py]: Yes, that's the check it is referring to.

File: core/src/apps/management/backup_device.py [hunk 12]
```diff
@@ -20,12 +22,12 @@
     storage.device.set_unfinished_backup(True)
     storage.device.set_backed_up()
 
-    if mnemonic_type == mnemonic.TYPE_SLIP39:
-        await backup_slip39_wallet(ctx, mnemonic_secret)
-    elif mnemonic_type == mnemonic.TYPE_SLIP39_GROUP:
-        await backup_group_slip39_wallet(ctx, mnemonic_secret)
+    if mnemonic_type == BackupType.Slip39_Basic:
```
[the remark this answers is not in the document]
Comment (ciny) [reply core/src/apps/management/backup_device.py]: LGTM. Where shall we put the function?

File: core/src/apps/management/reset_device/__init__.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (tsusanka) [reply core/src/apps/management/reset_device/__init__.py]: Pushed.

[the remark this answers is not in the document]
Comment (tsusanka) [reply core/src/apps/management/reset_device/__init__.py]: https://github.com/trezor/trezor-firmware/pull/518/commits/6008f4e234e1cb08676a87734083c988868766de @matejcik mark as resolved after inspection.

File: core/src/apps/management/recovery_device/homescreen.py [hunk 13]
```diff
@@ -48,33 +46,77 @@
     in_progress = storage.recovery.is_in_progress()
     word_count = storage.recovery.get_word_count()
     dry_run = storage.recovery.is_dry_run()
+    backup_type = storage.recovery.get_backup_type()
 
     if not in_progress:  # invalid and inconsistent state
```
[the remark this answers is not in the document]
Comment (tsusanka) [reply core/src/apps/management/recovery_device/homescreen.py]: Well, it makes sure the `_continue_recovery_process` is not called in case there is no recovery in progress. This should not happen. Maybe it's too much?

File: core/src/apps/management/reset_device/__init__.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (matejcik) [reply core/src/apps/management/reset_device/__init__.py]: the BIP39 branch could have been an `elif` ;) but that's not important

File: core/src/apps/management/backup_device.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (ciny) [reply core/src/apps/management/backup_device.py]: fixed. the function is in reset_device and it's imported into backup_device from there.

File: core/src/apps/management/recovery_device/recover.py [hunk 14]
```diff
@@ -24,47 +21,62 @@ def process_bip39(words: str) -> bytes:
     return words.encode()
```
[the remark this answers is not in the document]
Comment (ciny) [reply core/src/apps/management/recovery_device/recover.py]: fixed

File: core/src/apps/management/recovery_device/homescreen.py [hunk 15]
```diff
@@ -48,33 +46,77 @@
     in_progress = storage.recovery.is_in_progress()
     word_count = storage.recovery.get_word_count()
     dry_run = storage.recovery.is_dry_run()
+    backup_type = storage.recovery.get_backup_type()
```
[the remark this answers is not in the document]
Comment (tsusanka) [reply core/src/apps/management/recovery_device/homescreen.py]: `word_count` is being used few lines above:

```
    if not word_count:  # the first run, prompt word count from the user
        word_count = await _request_and_store_word_count(ctx, dry_run)
```

I didn't want to inline it, because the function was quite long before. That being said, it might be more readable now. I have moved the `backup_type` for now and left the rest, but if you want, feel free to inline the code.

File: core/src/apps/management/recovery_device/layout.py [hunk 16]
```diff
@@ -72,41 +69,60 @@
...
-                        if share_list[i] == word:
-                            raise ShareAlreadyAddedError()
-            else:
-                # in case of advanced shamir recovery we only check 2 words
-                if i < 2:
-                    share_list = mnemonics[0].split(" ")
-                    if share_list[i] != word:
-                        raise IdentifierMismatchError()
+        # only if we know the backup type we can perform additional checks
+        if backup_type and not await check_mnemonic_validity(ctx, i, word, backup_type):
```
[the remark this answers is not in the document]
Comment (ciny) [reply core/src/apps/management/recovery_device/layout.py]: fixed to 
backup_type is not None and ...

[the remark this answers is not in the document]
Comment (matejcik) [reply core/src/apps/management/recovery_device/layout.py]: ISTM this will fail with BIP39 now? or will it not get called?

File: core/src/apps/management/recovery_device/homescreen.py [hunk 17]
```diff
@@ -48,33 +46,77 @@
...
+
     digest_input = sha256(secret).digest()
     stored = mnemonic.get_secret()
     digest_stored = sha256(stored).digest()
     result = utils.consteq(digest_stored, digest_input)
 
+    is_slip39 = backup_type in (BackupType.Slip39_Basic, BackupType.Slip39_Advanced)
     # Check that the identifier and iteration exponent match as well
-    if mnemonic_type == mnemonic.TYPE_SLIP39:
+    if is_slip39:
```
[the remark this answers is not in the document]
Comment (ciny) [reply core/src/apps/management/recovery_device/homescreen.py]: I checked with @andrewkozlik and he said it wasn't necessary.

File: core/src/apps/management/recovery_device/homescreen.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (matejcik) [reply core/src/apps/management/recovery_device/homescreen.py]: `_continue_recovery_progress` is almost-directly called from `recovery_homescreen`, which only boots with `_boot_recovery`, which is only selected [`if recovery.is_in_progress()`](https://github.com/trezor/trezor-firmware/blob/iss442-supershamir-refactor/core/src/main.py#L91)

There should be a rationale for the check in a comment, or the check shouldn't be here. Otherwise it looks out of place; it's not clear from the context that this should be checked, and that this is the place to do it.

[the remark this answers is not in the document]
Comment (tsusanka) [reply core/src/apps/management/recovery_device/homescreen.py]: The idea was not to call `_continue_recovery_progress` from some wrong place. But I agree, removed.

File: core/src/apps/management/recovery_device/homescreen.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (matejcik) [reply core/src/apps/management/recovery_device/homescreen.py]: i played around with the code a little and decided to inline it, it makes more sense than keeping it separate

**Comment by matejcik** [comment]
I'm getting confused, so someone with a clearer head please take a look at the `backup_type` usage starting in `_continue_recovery_process`, and how it's propagated to the subroutines

what seems wrong to me is that there is a strong overlap of `backup_type` and `word_count` (through `is_slip39_word_count`). the values of `word_count`, `backup_type` and `is_slip39` should be kept in better sync than they are now.

File: core/src/apps/management/recovery_device/homescreen.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (tsusanka) [reply core/src/apps/management/recovery_device/homescreen.py]: I've looked at it shortly and it looks good to me! I lost perspective after I've move it 10 times back and forth :)

**Comment by tsusanka** [comment]
You are right there is an overlap, but I'm not sure what exactly do you mean. From word count we can infer only BIP-39 vs SLIP-39, but not Basic vs Advanced. That's why there is backup_type which is set after first mnemonic is entered - that's when we can find out. And is_slip39 as a bit of shortcut to not calling `is_slip39_word_count` repeatedly.

Let's have a look at this in real life.

File: core/src/apps/management/recovery_device/layout.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (tsusanka) [reply core/src/apps/management/recovery_device/layout.py]: I believe you are correct, because `check_slip39_mnemonic_validity` requires some previous mnemonics to be present. I've made it more explicit in  d28e69c79db047318e2a2514ffb867bebe0d4682

**Comment by tsusanka** [comment]
Closing in favour of  #539
